### PR TITLE
Better tracing alignment

### LINKED ISSUE #6523: Trace transaction - Investigate gasCost difference
### Description
Some differences in the Transaction Tracing:

On Mainnet the trace for the following transaction has a different gasCost from Geth and Nethermind
`0xd394531edb1a6b26c66d1f4e25f7367ceed116e826ef8e09d1b4b5017f40837f`

### geth
```
{
  "pc": 2029,
  "op": "STATICCALL",
  "gas": 2793331,
  "gasCost": 2749726,
  "depth": 1,
  "stack": [
    "0xc9567bf9",
    "0x242",
    "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
    "0xc45a0155",
    "0x84",
    "0x20",
    "0x80",
    "0x4",
    "0x80",
    "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
    "0x2a9f73"
  ]
}
```

### besu
```
{
  "pc": 2029,
  "op": "STATICCALL",
  "gas": 2793331,
  "gasCost": 2600,
  "depth": 1,
  "stack": [
    "00000000000000000000000000000000000000000000000000000000c9567bf9",
    "0000000000000000000000000000000000000000000000000000000000000242",
    "0000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488d",
    "00000000000000000000000000000000000000000000000000000000c45a0155",
    "0000000000000000000000000000000000000000000000000000000000000084",
    "0000000000000000000000000000000000000000000000000000000000000020",
    "0000000000000000000000000000000000000000000000000000000000000080",
    "0000000000000000000000000000000000000000000000000000000000000004",
    "0000000000000000000000000000000000000000000000000000000000000080",
    "0000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488d",
    "00000000000000000000000000000000000000000000000000000000002a9f73"
  ],
  "memory": null,
  "storage": {
    "0000000000000000000000000000000000000000000000000000000000000012": "0000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488d",
    "22ccc9016ce6aab1d473ca35a017c16a1e8e8d573fffb5e670ecaa2ce7bfeb88": "0000000000000000000000000000000000000000000000005fc1b97136320000"
  },
  "reason": null
}
```

### nethermind
```
{
  "pc": 2029,
  "op": "STATICCALL",
  "gas": 2793331,
  "gasCost": 2749726,
  "depth": 1,
  "error": null,
  "stack": [
    "0xc9567bf9",
    "0x242",
    "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
    "0xc45a0155",
    "0x84",
    "0x20",
    "0x80",
    "0x4",
    "0x80",
    "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
    "0x2a9f73"
  ],
  "storage": {
    "0000000000000000000000000000000000000000000000000000000000000012": "0000000000000000000000000000000000000000000000000000000000000012",
    "22ccc9016ce6aab1d473ca35a017c16a1e8e8d573fffb5e670ecaa2ce7bfeb88": "22ccc9016ce6aab1d473ca35a017c16a1e8e8d573fffb5e670ecaa2ce7bfeb88"
  }
}
```

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Update tracing and evm tool
* Intrinsic gas is optional in EVMTool
* For call series, also charge the gas given to the next call level

## Fixed Issue(s)

Fixes #6523

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/tracing/StandardJsonTracer.java [hunk 1]
```diff
@@ -174,6 +175,10 @@ public void tracePostExecution(
     }
     final int opcode = currentOp.getOpcode();
     final Bytes returnData = messageFrame.getReturnData();
+    long thisGasCost = executeResult.getGasCost();
+    if (currentOp instanceof AbstractCallOperation) {
+      thisGasCost += messageFrame.getMessageFrameStack().getFirst().getRemainingGas();
+    }
```
Comment (Gabriel-Trintinalia) [inline evm/src/main/java/org/hyperledger/besu/evm/tracing/StandardJsonTracer.java:181]: @shemnon  Should we do the same change in `DebugOperationTracer`?
```
            operationResult.getGasCost() == 0
                ? OptionalLong.empty()
                : OptionalLong.of(operationResult.getGasCost()),
```

Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/tracing/StandardJsonTracer.java:181]: It required a bit more blocking and tacking to make sure it didn't leak into the `trace_` series of calls, which do actual consumed gas separately.  However it is done.

File: ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/EvmToolCommand.java [hunk 2]
```diff
@@ -395,6 +387,33 @@ public void run() {
         var contractAccount = updater.getOrCreate(contract);
         contractAccount.setCode(codeBytes);
 
+        final Set<Address> addressList = new BytesTrieSet<>(Address.SIZE);
+        addressList.add(sender);
+        addressList.add(contract);
```
Comment (Gabriel-Trintinalia) [inline ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/EvmToolCommand.java:392]: What does this change mean in this PR?

**Review by Gabriel-Trintinalia — APPROVED** [review]
LGTM, Would you be able to add some comments on how `additionalCallGas` works  either in the `AbstractTraceCall` or in the `DebugOperationTracer` and/or when it makes sense to use it? It could help future development to use this flag correctly.

**Review by ahamlat — APPROVED** [review]
LGTM. From what I see, this pull request makes two separate improvements. The change in tracing affects only calls frames in debug trace RPC methods, not "Parity" style RPC tracing calls. I'm assuming this difference is more about the design or specification choice, correct?

File: ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/EvmToolCommand.java (hunk 2 above)
Comment (shemnon) [reply ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/EvmToolCommand.java:392]: EIP-2929 says the sender and contract address are pre-warmed.  The transaction handler does that, but the evm tool has it's own bootstrap logic and needs to pre-warm them.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/DebugOperationTracerTest.java [hunk 3]
```diff
@@ -129,41 +134,55 @@ void shouldRecordMemoryWhenEnabled() {
...
+    assertThat(traceFrame.getGasCost()).isEqualTo(OptionalLong.of(20));
+  }
+
+  @Test
+  void shouldAddGasWhenEnabled() {
+    final TraceFrame traceFrame =
+        traceFrame(validCallFrame(), new TraceOptions(false, false, false), true);
+    assertThat(traceFrame.getGasCost()).isEqualTo(OptionalLong.of(1020L));
+  }
+
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/DebugOperationTracerTest.java:178]: test added.  validMessageFrame with opposite flags, asserting it's the same.

**Review by ahamlat — APPROVED** [review]

**Pull request merged**